### ISSUE TITLE
Number of Items in Section Calendar crash

### ISSUE THREAD

**Issue by mmcatcd** [patchthecode/JTAppleCalendar#840]
Version: 7.1.5

Hi, I get the following error when calling calendar.numberOfItems(inSection: ) in either `viewDidLoad` or in `viewDidAppear`.

`Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'UICollectionView received layout attributes for a cell with an index path that does not exist: <NSIndexPath: 0x1d403d980> {length = 2, path = 0 - 0}'
*** First throw call stack:
(0x183a5ed8c 0x182c185ec 0x183a5ebf8 0x18444efa0 0x18d943440 0x18d7a49e8 0x18d7a36b4 0x104731270 0x10473182c 0x18d6726f4 0x187be7e54 0x187bebfe4 0x187b586c8 0x187b801b0 0x187b81030 0x183a06910 0x183a04238 0x183a04884 0x183924da8 0x18590a020 0x18d944758 0x1041658c4 0x1833b5fc0)
libc++abi.dylib: terminating with uncaught exception of type NSException`

The calendar that I've tested it on is supposed to have 64 sections and the error occurs no matter which section I try to get the number of items from. It appears that at the point that the function is called, the calendar is not loaded and as a result has 0 sections so it is unable to get the number of items.

My configuration:

```swift
func configureCalendar(_ calendar: JTAppleCalendarView) -> ConfigurationParameters {
        var myCalendar = Calendar(identifier: .gregorian)
        myCalendar.timeZone = TimeZone(secondsFromGMT: 0)!
        
        let parameters = ConfigurationParameters(
            startDate: startDate,
            endDate: endDate,
            numberOfRows: 1,
            calendar: myCalendar,
            generateInDates: .forFirstMonthOnly,
            generateOutDates: .off,
            firstDayOfWeek: .sunday)
        
        //Select current date by default
        calendar.selectDates([Date()])
        calendar.scrollToDate(Date())
        
        return parameters
    }
```

My delegate:

```swift
func calendar(_ calendar: JTAppleCalendarView, cellForItemAt date: Date, cellState: CellState, indexPath: IndexPath) -> JTAppleCell {
        let cell = calendar.dequeueReusableJTAppleCell(withReuseIdentifier: "ScheduleViewCell", for: indexPath) as! ScheduleViewCell
        configureCell(cell: cell, cellState: cellState, date: date)
        
        return cell
    }
    
    func calendar(_ calendar: JTAppleCalendarView, willDisplay cell: JTAppleCell, forItemAt date: Date, cellState: CellState, indexPath: IndexPath) {
        let customCell = cell as! ScheduleViewCell
        configureCell(cell: customCell, cellState: cellState, date: date)
    }
    
    func configureCell(cell: ScheduleViewCell, cellState: CellState, date: Date) {
        cell.dateLabel.text = cellState.text
        
        // Fixes problem with UICollectionView reusing cells
        if cellState.isSelected {
            selectCell(cell: cell, date: date)
        } else {
            deselectCell(cell: cell)
        }
    }
    
    func calendar(_ calendar: JTAppleCalendarView, didSelectDate date: Date, cell: JTAppleCell?, cellState: CellState) {
        guard let validCell = cell as? ScheduleViewCell else { return }
        
        collectionView.scrollToItem(at: calendar.indexPath(for: cell!)!, at: .right, animated: true)
        
        selectCell(cell: validCell, date: date)
        
        // Change date to select data from
        currentDate = dateFormatter.string(from: date)
    }
    
    func calendar(_ calendar: JTAppleCalendarView, didDeselectDate date: Date, cell: JTAppleCell?, cellState: CellState) {
        guard let validCell = cell as? ScheduleViewCell else { return }
        deselectCell(cell: validCell)
    }
    
    func selectCell(cell: ScheduleViewCell, date: Date) {
        cell.selectView.isHidden = false
        cell.dateLabel.textColor = UIColor.black
        dateLabel.text = dateLabelFormatter.string(from: date)
    }
    
    func deselectCell(cell: ScheduleViewCell) {
        cell.selectView.isHidden = true
        cell.dateLabel.textColor = UIColor.white
    }
```

I'm baffled as to what may be the issue. I was running into an issue where the value being returned by numberOfSections was always zero. I saw #775 and that he had noticed that the calendar hadn't loaded in the `viewWillAppear` but did load in the `viewDidAppear`. It appears to me that even in the `viewDidAppear` the calendar has not been loaded.

Is this an expected behaviour and if so is there a method that I can call where it is expected that the calendar will be loaded?

Any help would be greatly appreciated on this issue. Let me know if you need any more info.

**Comment by patchthecode** [comment]
Can you zip a sample project that has this error and drag it into this chat box?

**Comment by patchthecode** [comment]
will close this issue issue until user responds.
added tag of awaiting user input.

**Comment by mmcatcd** [comment]
My apologies for the late reply on this. I managed to solve the issue by adding: ```calendar.numberOfItems(inSection:)``` to the main DispatchQueue from within the ```viewDidAppear```. I'm not quite sure why this seems to work but it appears that because the ```viewDidAppear``` runs on the main thread to begin with, anything that I add to the queue from within it will execute after the ```viewDidAppear``` method.

The application I'm having the problem with is relatively complex as it has the calendar and a horizontally paging collection view underneath, with each cell containing a table view to display a daily schedule for the selected date on the calendar. I was not able to reproduce the problem with simpler examples. There could be something in the code that I'm unable to find causing a delay in the load time and adding it to the Dispatch Queue gives it just enough time to load in.

Anyway it seems to work for now. The code I used in the ```ViewDidAppear``` was:

```
override func viewDidAppear(_ animated: Bool) {
        DispatchQueue.main.async {
            print(self.calendar.numberOfItems(inSection: 0))
        }
    }

```
